sources/packages/backend/apps/api/src/services/application-restriction-bypass/application-restriction-bypass.service.ts

```diff
@@ -49,4 +82,380 @@ export class ApplicationRestrictionBypassService extends RecordDataModelService<
       order: { createdAt: "DESC" },
     });
   }
+
+  /**
+   * Gets an application restriction bypass by id.
+   * @param id id of the application restriction bypass to retrieve.
+   * @returns application restriction bypass.
+   */
+  getApplicationRestrictionBypass(
+    id: number,
+  ): Promise<ApplicationRestrictionBypass> {
+    return this.repo.findOne({
+      select: {
+        id: true,
+        isActive: true,
+        studentRestriction: {
+          id: true,
+          restriction: {
+            restrictionCode: true,
+          },
+        },
+        creationNote: true as unknown,
+        removalNote: true as unknown,
+        bypassCreatedBy: true as unknown,
+        bypassCreatedDate: true,
+        bypassRemovedBy: true as unknown,
+        bypassRemovedDate: true,
+        bypassBehavior: true,
+      },
+      relations: {
+        studentRestriction: { restriction: true },
+        creationNote: true,
+        removalNote: true,
+        bypassCreatedBy: true,
+        bypassRemovedBy: true,
+      },
+      where: {
+        id,
+        application: {
+          applicationStatus: Not(
+            In([
+              ApplicationStatus.Draft,
+              ApplicationStatus.Cancelled,
+              ApplicationStatus.Overwritten,
+            ]),
+          ),
+        },
+      },
+    });
+  }
+
+  /**
+   * Gets all available student restrictions to bypass for a given application.
+   * @param applicationId application id.
+   * @returns list of application restriction bypass options.
+   */
+  async getAvailableStudentRestrictionsToBypass(
+    applicationId: number,
+  ): Promise<StudentRestriction[]> {
+    return await this.studentRestrictionRepo
+      .createQueryBuilder("studentRestriction")
+      .select([
+        "studentRestriction.id",
+        "restriction.restrictionCode",
+        "studentRestriction.createdAt",
+      ])
+      .innerJoin("studentRestriction.restriction", "restriction")
+      .innerJoin("studentRestriction.application", "application")
+      .innerJoin("application.currentAssessment", "currentAssessment")
+      .innerJoin("currentAssessment.offering", "offering")
+      .where("studentRestriction.application = :applicationId", {
+        applicationId,
+      })
+      // Active application restriction bypass.
+      .andWhere((qb) => {
+        const subQuery = qb
+          .subQuery()
+          .select("applicationRestrictionBypass.id")
+          .from(ApplicationRestrictionBypass, "applicationRestrictionBypass")
+          .where("applicationRestrictionBypass.application = :applicationId", {
+            applicationId,
+          })
+          .andWhere(
+            "applicationRestrictionBypass.studentRestriction = studentRestriction.id",
+          )
+          .andWhere("applicationRestrictionBypass.isActive = true");
+        return `NOT EXISTS (${subQuery.getQuery()})`;
+      })
+      // Restriction action type condition.
+      .andWhere((qb) => {
```

Comment:
Please correct me if I am missing something, but knowing before query execution if the application is part-time/full-time would dramatically simplify this entire where to a simple condition. Does it make sense?
Also, the number of active student restrictions at any given time would be probably 1 to few, which means that it could be even filtered in memory to avoid a complex SQL query.

Also, the query relies on the fact the student restriction will be associated with an application when it may not.